qpsolvers/solvers/scs_.py

```diff
@@ -105,40 +106,59 @@ def scs_solve_qp(
-    problem. We lower it to ``1e-7`` and switch ``use_indirect=True`` so that
-    SCS behaves closer to the other solvers on this example.
+    Keyword arguments are forwarded as is to SCS. For instance, you can call
+    ``scs_solve_qp(P, q, G, h,normalize=True)``. Solver settings for SCS are
+    described `here <https://www.cvxgrp.org/scs/api/settings.html#settings>`_.
 
-    All other keyword arguments are forwarded as is to SCS. For instance, you
-    can call ``scs_solve_qp(P, q, G, h, use_indirect=True, normalize=True)``.
-    See the solver documentation for details.
+    As of SCS 3.0.1, the default feasibility tolerances are set ``1e-4``,
+    resulting in larger inequality constraint violations than with other
+    solvers on the README and unit test problems. We lower them to ``1e-9`` so
+    that SCS behaves closer to the other solvers. If you don't need that much
+    precision, increase them for better performance.
     """
-    if initvals is not None:
-        warn("note that warm-start values ignored by this wrapper")
-    c_socp, G_socp, h_socp, dims = convert_to_socp(P, q, G, h)
+    if isinstance(P, ndarray):
+        warn_about_sparse_conversion("P")
+        P = sparse.csc_matrix(P)
+    if isinstance(G, ndarray):
+        warn_about_sparse_conversion("G")
+        G = sparse.csc_matrix(G)
+    if isinstance(A, ndarray):
+        warn_about_sparse_conversion("A")
+        A = sparse.csc_matrix(A)
     kwargs.update(
-        {"eps": eps, "use_indirect": use_indirect, "verbose": verbose}
+        {
+            "eps_abs": eps_abs,
+            "eps_rel": eps_rel,
+            "verbose": verbose,
+        }
     )
+    data = {"P": P, "c": q}
+    cone = {}
+    if initvals is not None:
+        data["x"] = initvals
     if A is not None and b is not None:
-        dims["f"] = A.shape[0]  # number of equality constraints
-        A_socp = sparse.hstack(
-            [A, sparse.csc_matrix((A.shape[0], 1))], format="csc"
-        )
-        A_scs = sparse.vstack([A_socp, G_socp], format="csc")
-        b_scs = hstack([b, h_socp])
-        data = {"A": A_scs, "b": b_scs, "c": c_socp}
-        solution = solve(data, dims, **kwargs)
-    else:
-        data = {"A": G_socp, "b": h_socp, "c": c_socp}
-        solution = solve(data, dims, **kwargs)
-    status_val = solution["info"]["statusVal"]
+        if G is not None and h is not None:
+            data["A"] = sparse.vstack([A, G], format="csc")
+            data["b"] = hstack([b, h])
+            cone["z"] = b.shape[0]  # zero cone
+            cone["l"] = h.shape[0]  # positive orthant
+        else:  # A is not None and b is not None
+            data["A"] = A
+            data["b"] = b
+            cone["z"] = b.shape[0]  # zero cone
+    elif G is not None and h is not None:
+        data["A"] = G
+        data["b"] = h
+        cone["l"] = h.shape[0]  # positive orthant
+    else:  # no constraint
+        return sparse.linalg.lsqr(P, -q)[0]
```

Comment:
The only issue is if `q` has a component in the null space of `P`, in which case the problem is unbounded below, since we can set `Px = 0` and `q'x` to be as negative as we want. We can check this using something like:

```
x_star = sparse.linalg.lsqr(P, -q)[0]
if np.linalg.norm(P @ x_star - q) > 1e-9: 
  # return something to say this problem is unbounded
else:
  return x_star
```

Reply:
Right! Oh my, actually most functions assume that `P` is PSD :see_no_evil: I've created https://github.com/stephane-caron/qpsolvers/issues/51 to make sure this gets covered in the future.

Added this check in 469da9bac33477225bc2f327c108acb7a055641d :ok_hand: